includes/ProductSync/ProductValidator.php

```diff
@@ -0,0 +1,203 @@
+	 */
+	protected function validate_sync_enabled_globally() {
+		if ( $this->integration->is_product_sync_enabled() ) {
+			throw new ProductExcludedException( 'Product sync is globally disabled.' );
+		}
+	}
+
+	/**
+	 * Check whether the product's status excludes it from sync.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	protected function validate_product_status() {
+		$product = $this->product_parent ?: $this->product;
+
+		if ( 'publish' !== $product->get_status() ) {
+			throw new ProductExcludedException( 'Product is not published.' );
+		}
+	}
+
+	/**
+	 * Check whether the product should be excluded due to being out of stock.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	protected function validate_product_stock_status() {
+		if ( 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' ) && ! $this->product->is_in_stock() ) {
+			throw new ProductExcludedException( 'Product must be in stock.' );
+		}
+	}
+
+	/**
+	 * Check whether the product's visibility excludes it from sync.
+	 *
+	 * Products are excluded if they are hidden from the store catalog or from search results.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	protected function validate_product_visibility() {
+		$product = $this->product_parent ?: $this->product;
+
+		if ( 'visible' !== $product->get_catalog_visibility() ) {
+			throw new ProductExcludedException( 'Product is hidden from catalog and search.' );
+		}
+	}
+
+	/**
+	 * Check whether the product's categories or tags exclude it from sync.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	public function validate_product_categories_and_tags() {
+		$product = $this->product_parent ?: $this->product;
+
+		$excluded_categories = $this->integration->get_excluded_product_category_ids();
+		if ( $excluded_categories ) {
+			if ( ! empty( array_intersect( $product->get_category_ids(), $excluded_categories ) ) ) {
+				throw new ProductExcludedException( 'Product excluded because of categories.' );
+			}
+		}
+
+		$excluded_tags = $this->integration->get_excluded_product_tag_ids();
+		if ( $excluded_tags ) {
+			if ( ! empty( array_intersect( $product->get_tag_ids(), $excluded_tags ) ) ) {
+				throw new ProductExcludedException( 'Product excluded because of tags.' );
+			}
+		}
+	}
+
+	/**
+	 * Validate if the product is excluded from at the "product level" (product meta value).
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	public function validate_product_sync_field() {
+		$invalid_exception = new ProductExcludedException( 'Sync disabled in product field.' );
+
+		if ( $this->product->is_type( 'variable' ) ) {
+			foreach ( $this->product->get_children() as $child_id ) {
+				$child_product = wc_get_product( $child_id );
+				if ( $child_product && 'no' !== $child_product->get_meta( self::SYNC_ENABLED_META_KEY ) ) {
+					break;
```

Comment:
```suggestion
					return;
```
I think that the logic should check for at least one child with sync enabled. I think that this was the intention.
With the `break;`  the code is meaningless because we will always reach `throw`.

Reply:
Oops my mistake. Fixed in c90ffca010ad6c8dd29e7e18a49fe01e9b029768.